### PR TITLE
Adding EIP for described data

### PR DESCRIPTION
**Pull request by ricmoo**
Adding initial draft for described data to allocate an EIP so a discussion-to URL can be added in the issues of the EIP repo.

**Abstract**

Human-readable descriptions for machine executable operations,
described in higher level machine readable data, so that wallets
can provide meaningful feedback to the user describing the
action the user is about to perform.

### CODE REVIEW

File: EIPS/eip-3224.md
Comment (MicahZoltu) [inline EIPS/eip-3224.md:57]: Most of these names don't align with parameter names in the Solidity below, which makes it really difficult to tell what is being talked about.  Recommend following a consistent naming scheme throughout the document.

Comment (ricmoo) [reply EIPS/eip-3224.md:57]: You mean I should rename the signature to:

```
function eipXXXDescribe(bytes describer_input) view returns (string description_string, bytes described_data);
```

?

I can make that change now.

File: EIPS/eip-3224.md
Comment (MicahZoltu) [inline EIPS/eip-3224.md:81]: Displayed where?  Also, in general `should` is a red flag in a standard as anything that is a `should` or `may` cannot be enforced for compatibility.  This is especially true in this case where I *think* you are trying to tell user interfaces what they should do, which is more appropriate for an opinion piece like a blog post, tutorial, best practices document, patterns book, etc.

File: EIPS/eip-3224.md
Comment (MicahZoltu) [inline EIPS/eip-3224.md:76]: Maybe I'm misunderstanding this specification (very possible), but it seems like none of this needs to be part of the standard.  The specification I believe can be simplified to just an ERC that defines an `eipXXXDescribe` function which takes in a byte array and returns a string (and maybe that data parameter, it is not clear to me what exactly that does or is for yet).

File: EIPS/eip-3224.md
Comment (MicahZoltu) [inline EIPS/eip-3224.md:97]: Above you are specifying a contract interface (ERC) and here you are specifying new JSON-RPC methods (Interface).  This should be split into two EIPs, where perhaps the JSON-RPC one has a dependency on the contract interface one.

File: EIPS/eip-3224.md
Comment (MicahZoltu) [inline EIPS/eip-3224.md:123]: Note: EIP-191 is still just a draft, so this EIP will get blocked by that unless you can move it forward.

File: EIPS/eip-3224.md
Comment (ricmoo) [reply EIPS/eip-3224.md:76]: It requires returning a string and a bytes. The bytes passed in are a higher level description of the bytes being asked to be generated. A good example of this is the ENS commit example.

The input data describes what is wanted, the string is shown to the user (e.g. in metamask) and the output data is the data to use in a transaction.

File: EIPS/eip-3224.md
Comment (MicahZoltu) [inline EIPS/eip-3224.md:122]: Why does the described data have to be signed?  What value does that add?  Who are you trying to prove this to that cannot just run the describer again?

Comment (ricmoo) [reply EIPS/eip-3224.md:122]: As an example, OpenSea today has you sign a hash, which is submitted to a contract. That hash would not change, and would still be what is required to be signed. What does change is what the UI shows you; instead of "sign this hash: 0xsometing", it would show "Do you agree to place item 1234 for sale at a price of 1.23 ether". Upon accepting this, the wallet would continue signing the data, submitting it to OpenSea's servers, at this point nothing much changes.

File: EIPS/eip-3224.md
Comment (ricmoo) [reply EIPS/eip-3224.md:123]: Interesting... @Arachnid, is there any reason not to move this forward?

File: EIPS/eip-3224.md
Comment (ricmoo) [reply EIPS/eip-3224.md:81]: I guess the reason for the *should* is that not all UI's are interactive or have a display.

File: EIPS/eip-3224.md
Comment (ricmoo) [reply EIPS/eip-3224.md]: I've updated this section.

Comment (MicahZoltu) [reply EIPS/eip-3224.md]: Yes, though I still don't know what `describer` is.

Comment (ricmoo) [reply EIPS/eip-3224.md]: The describer is just the contract itself. I can just change the name to contract? Originally this was more abstract, but we decided it made more sense to reduce the scope to a contract doing this itself (as a delegatecall can be used to outsource to another contract rather than returning a separated address to operate as the describer).

File: EIPS/eip-3224.md
Comment (MicahZoltu) [inline EIPS/eip-3224.md:66]: You have two open parenthesis and one close parenthesis.

File: EIPS/eip-3224.md
Comment (MicahZoltu) [inline EIPS/eip-3224.md:72]: Consider splitting out `describeTransaction` and `describeMessage` into two different methods.

If you do that, you can split them into two different EIPs.  This would allow a contract to implement one or the other, which can improve discoverability and make it easier to implement.

File: EIPS/eip-3224.md
Comment (MicahZoltu) [inline EIPS/eip-3224.md:76]: Consider instead asserting that these **MUST NOT** be referenced in the `describe` function.  Alternatively, you could just say they **SHOULD NOT** be referenced and then mention in the rationale section that while in some obscure cases they may be necessary, care should be taken not to assume that signers will be signing against the most recent block (since that is effectively impossible due to the nature of distributed systems).

File: EIPS/eip-3224.md
Comment (MicahZoltu) [reply EIPS/eip-3224.md:82]: This paragraph feels out of scope for an ERC.  UIs can do whatever they want once they have completed the interaction with the contract.  I do agree that what you have described is a Good Idea, but EIPs are for standardization, not Good Ideas.  😄

File: EIPS/eip-3224.md
Comment (MicahZoltu) [inline EIPS/eip-3224.md:466]:
```suggestion
```

EIPs should be written toward their final form, and this section isn't a valid EIP section so should be removed.  If you want, you can mark this PR as a review so people don't start reading it, and then you can delete this before submitting to be merged as a draft.  Alternatively, consider moving this content to something like HackMD or the discussions-to link.

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
Only real blocker is the footer, which should be deleted.  If you aren't ready to merge this as a draft you can mark the PR as a draft PR and then editors won't look at it.

The rest of the feedback I encourage reviewing.  We also discussed in Discord a lot of it and I think we *generally* are in agreement for most things, but that is largely unrelated to my editorial duties.

File: EIPS/eip-3224.md
Comment (MicahZoltu) [inline EIPS/eip-3224.md:247]: Generally, we should try to avoid linking to external resources as the links will almost certainly break with time and editors do not have the manpower to maintain them.  It is acceptable to link to published (non-draft) EIPs, but for this section I recommend leaving out the links and specific products and instead just describing the alternative solutions briefly.

Comment (MicahZoltu) [reply EIPS/eip-3224.md:247]: (same goes for the NatSpec link above)

**Comment by ricmoo** [comment]
Thanks @MicahZoltu! I’ll get these updates made today. :)

File: EIPS/eip-3224.md
Comment (ricmoo) [reply EIPS/eip-3224.md:82]: I've removed this and simple mentioned that the method may revert and that if so, the signing must be aborted.

File: EIPS/eip-3224.md
Comment (ricmoo) [reply EIPS/eip-3224.md:98]: I will bring up splitting this EIP up on the discussion-to link, to help figure out exactly where the lines should be drawn. :)

File: EIPS/eip-3224.md
Comment (ricmoo) [reply EIPS/eip-3224.md:72]: And I will include a reference to EIP-165 for discovery.

File: EIPS/eip-3224.md
Comment (ricmoo) [reply EIPS/eip-3224.md:76]: I think there are common (may not common, but not uncommon) case where they are needed, which I'll flesh out in the rationale. Even something as simple as CryptoKitties could use it for the birth alarm clock, although it may make sense to discourage alarm clocks...

**Comment by ricmoo** [comment]
It took longer to get to than expected. But I've addressed the reviews above.

I'll also start two discussions on the discussion-to link; localization and splitting the EIP up. @Arachnid and I have been having some conversations regarding this, so it would be good to get them out for others to comment on.

File: EIPS/eip-3224.md
Comment (ricmoo) [reply EIPS/eip-3224.md:119]: I think for now this is fine, but will need to be addressed later. But we certainly don't want to define some new standard that's not already in the works for something like this. EIP-191 is already widely used.

**Comment by ricmoo** [comment]
I've resolved all the conversations. Let me know if there is something else that is needed.

Thanks! :)

File: EIPS/eip-3224.md
Comment (MicahZoltu) [inline EIPS/eip-3224.md:410]: External links should be avoided whenever possible, as they tend to 404 over time and we don't have the manpower to maintain dead links.  I think most of these are easily searchable, so no need to link to them (naming them is fine).

**Review by MicahZoltu — APPROVED** [review]

**Comment by Chucky-ux** [comment]
I know nothing how it works the coupling coins back to single coins.
Thanks for helping  me with all this issues!

Gesendet von Yahoo Mail auf Android 
 
  Am Mi., Feb. 17, 2021 at 11:34 schrieb Micah Zoltu<notifications@github.com>:   


Merged #3224 into master.

**Pull request merged**
